### PR TITLE
Plain text diff

### LINKED ISSUE #352: Default diff viewer
Hey,

Work planned in #331 is awesome. I think it will be nice to provide directly within this extension a default diff viewer (supporting for instance only raw text diff).

The goals:
- Provide a default diff viewer usefull enough for raw text diff
- Provide an example how to implement the API resulting of #331 for diff viewer developers

Resources:
The [merge addon](https://codemirror.net/doc/manual.html#addon_merge) of CodeMirror could be the basis for such default viewer. See the demo: https://codemirror.net/demo/merge.html
Another candidate could be the [monaco diff viewer](https://microsoft.github.io/monaco-editor/).

Note
I hope to have time to look into this in June. Is there a plan for #331 completion?

### LINKED ISSUE #446: Failed to fetch diff
Server responds: `HTTP 422: Unprocessable Entity (Invalid notebook: pipelines/base/index.ipynb)`

On my filesystem, I have files like this

```
/ (/home/jovyan)
|-- pipelines/ (my cloned git repo)
|---- base/
```

On request payload it shows as `"file_path":"pipelines/base/index.ipynb"`

I can commit/push/pull, but cannot view difference.

I tried to run request with postman and changing `file_path`, but even when I changed `file_path` to `/home/jovyan/pipelines/base/index.ipynb` or `base/index.ipynb` or `/base/index.ipynb` it always returns HTTP 422.

And I'm running this in docker-compose. base image is `jupyter/pyspark-notebook` with root + sudo enabled.

### PR DESCRIPTION
**Pull request by fcollonval**
fixes #352, fixes #446

This build on #388 but using CodeMirror instead of Monaco. The reason to switch to CodeMirror is to fit better with the visual identity of JupyterLab and nbdime that both use CodeMirror to show code. But the diff are highlighted following a style closer to Monaco than the default CodeMirror addon.

Here is a preview:

image

The remaining tasks:
- [x] Improve styling to inherit from JupyterLab codemirror styling so changing editor theme is propagate here
- [x] Add gutter + and -
- [x] Restructure `mergeview.ts`
- Open issues to link `codemirror` settings changed signal with diff widget to improve integration within jupyterlab

The diff/merge code comes from the merge addon of CodeMirror. This will allow to implement at low cost a plain text merge interface. 
The major modification to the original code is the usage of `diff-match-patch` npm package and the default view for the two panels case is to set the old text on the left panel and the current one in the right panel.

### CODE REVIEW

File: jupyterlab_git/git.py [hunk 1]
```diff
@@ -728,3 +728,53 @@ def _get_tag(self, current_path, commit_sha):
...
+                error.decode('utf-8'),
+                ' '.join(command)
+            ))
+
+    def less(self, filename, top_repo_path):
+        """
+        Execute less -FX <filename> command & return the result.
+        """
+        my_output = subprocess.check_output(
+            ["less", "-FX", filename], cwd=top_repo_path)
```
Comment (fcollonval) [inline jupyterlab_git/git.py]: This does not seems generic; on Windows it will fail. What it the goal here?

File: jupyterlab_git/git.py [hunk 2]
```diff
@@ -728,3 +728,53 @@ def _get_tag(self, current_path, commit_sha):
...
+                ' '.join(command)
+            ))
+
+    def get_content(self, filename, top_repo_path):
+        """
+        Get the file content of filename.
+        """
+        relative_repo = os.path.relpath(top_repo_path, self.root_dir)
+        model = self.contents_manager.get(path=os.path.join(relative_repo, filename))
+        return model.content
```
Comment (tomskikh) [inline jupyterlab_git/git.py]: `contents_manager.get()` returns dict (https://github.com/jupyter/notebook/blob/master/notebook/services/contents/filemanager.py#L417)
```suggestion
        return model['content']
```

Comment (fcollonval) [reply jupyterlab_git/git.py]: Thx for testing and the suggestion

**Comment by fcollonval** [comment]
@telamonian would you be willing to review this PR?

**Comment by telamonian** [comment]
ahh, I see you have another short PR for me, @fcollonval :) 

I actually already started going over this PR last weekend, so I'll get through it

File: src/components/diff/Diff.tsx [hunk 3]
```diff
@@ -14,17 +15,30 @@ const DIFF_PROVIDER_REGISTRY: { [key: string]: typeof React.Component } = {
...
- * This will be removed when "Plaintext" diffs are supported since that will be used
- * for cases where there is not a dedicated diff provider.
- *
  * @param path the file path
  */
 export function isDiffSupported(path: string): boolean {
-  return PathExt.extname(path).toLocaleLowerCase() in DIFF_PROVIDER_REGISTRY;
+  return (
+    PathExt.extname(path).toLocaleLowerCase() in DIFF_PROVIDER_REGISTRY ||
+    isText(path)
```
Comment (telamonian) [inline src/components/diff/Diff.tsx]: Even with `isText` added this is still not a great way to determine diff-ability. The problem is that it still misses many plaintext files named with weird or missing extensions that do handled correctly by eg `git diff`.

We can save this problem for a later PR, but I wanted to point it out here

File: src/components/diff/PlainTextDiff.tsx [hunk 4]
```diff
@@ -0,0 +1,158 @@
+import { ServerConnection } from '@jupyterlab/services';
+import { Mode } from '@jupyterlab/codemirror';
+import * as React from 'react';
+
+import { httpGitRequest } from '../../git';
+import { IDiffProps } from './Diff';
+import { getRefValue, IDiffContext } from './model';
+
+import { mergeView } from './mergeview';
```
Comment (telamonian) [inline src/components/diff/PlainTextDiff.tsx]: sort/alphabetize these imports

File: style/index.css [hunk 5]
```diff
@@ -3,5 +3,7 @@
 | Distributed under the terms of the Modified BSD License.
 |----------------------------------------------------------------------------*/
 
-@import 'diff.css';
 @import 'variables.css';
+@import 'diff-common.css';
+@import 'diff-text.css';
+@import 'diff-nb.css';
```
[the remark this answers is not in the document]
Comment (telamonian) [reply style/index.css]: This was so petty I just took care of it myself

File: src/components/diff/PlainTextDiff.tsx (hunk 4 above)
Comment (telamonian) [reply src/components/diff/PlainTextDiff.tsx]: took care of it

File: style/variables.css [hunk 6]
```diff
@@ -20,8 +20,8 @@
   --jp-icon-git-clone-disabled: url('./images/git-clone-icon-disabled.svg');
   --jp-icon-git-push-disabled: url('./images/git-clone-icon-disabled.svg');
   --jp-icon-git-pull-disabled: url('./images/git-clone-icon-disabled.svg');
-  --jp-git-diff-deleted-color: rgba(255, 193, 193, 0.5);
-  --jp-git-diff-added-color: rgba(201, 243, 194, 0.5);
+  --jp-git-diff-deleted-color: rgba(255, 0, 0, 0.2);
+  --jp-git-diff-added-color: rgba(155, 185, 85, 0.2);
```
[the remark this answers is not in the document]
Comment (fcollonval) [reply style/variables.css]: The difference can be seen here for the dark theme (new left - old right):

image

The new colors are the one used in VS code. But the one in need for improvement is the "string" color.

File: src/components/diff/Diff.tsx (hunk 3 above)
Comment (fcollonval) [reply src/components/diff/Diff.tsx]: Thanks for the creation of the related issue

**Comment by fcollonval** [comment]
Thanks for the review @telamonian 

As you have started creating follow-up issues, I'll merge this one. Feel free to open a new issue, if you think the diff theme for the dark them needs improvements.

> Are you the one who wrote `mergeview.ts`, or did you get from somewhere else? I'm mostly asking out of curiosity/being impressed. It's supposed to be a Typescript translation of https://github.com/codemirror/CodeMirror/blob/master/addon/merge/merge.js, right?

I have not much credit. I copied the original JS file you referenced and modified it to get a Typescript file without too many `ts-ignore` and without breaking its features.

**Comment by damianavila** [comment]
@fcollonval @telamonian this is on master but not released yet, am I right? in general, Is the milestone associated with the PRs/issues a reliable way to follow the *status quo*? Should I look into other documents besides the roadmap? Thanks!

**Pull request merged**